ops/jujuversion.py

```diff
@@ -0,0 +1,75 @@
+import re
+from functools import total_ordering
+
+
+@total_ordering
+class JujuVersion:
+
+    PATTERN = r'^(\d{1,9})\.(\d{1,9})((?:\.|-([a-z]+))(\d{1,9}))?(\.(\d{1,9}))?$'
+
+    def __init__(self, version):
+        m = re.match(self.PATTERN, version)
+        if not m:
+            raise RuntimeError(f"{version} is not a valid Juju version string.")
+
+        self.major = int(m.group(1))
```

Comment:
IMO, these are slightly nicer if you use the
(?P<name>\d{1,9}) syntax.
Then you can do:
m.group('name')

so something like:
```
PATTERN = r'^(\d{1,9})\.(\d{1,9})((?:\.|-([a-z]+))(\d{1,9}))?(\.(\d{1,9}))?$'
```
becomes
```
PATTERN = r'^(?P<major>\d{1,9})\.((?P<minor>\d{1,9})((?:\.|-(?P<tag>[a-z]+))(?P<patch>\d{1,9}))?(\.(?P<build>\d{1,9}))?$'
```

You can potentially use "m.groupdict()" to get all the named groups and then do:
patch = d['patch'], etc.

Reply:
OK, that would definitely be more readable.

I wanted to keep modifications at minimum compared to what I found in `juju/version` repo but this would be a good improvement.

Reply:
Just to elaborate on the last bit of John's comment, your `__init__` could be:

```python
    def __init__(self, version):
        m = self.PATTERN.match(version)
        if not m:
            raise RuntimeError(f"{version} is not a valid Juju version string.")
        d = m.groupdict()
        self.major = int(d['major'])
        self.minor = int(d['minor'])
        self.tag = d['tag'] or ''
        self.patch = int(d['patch'] or 0)
        self.build = int(d['build'] or 0)
```

Reply:
Thanks, this is much more concise than what I had.